### PR TITLE
[Flang][OpenMP][Lower] Add lowering support of OpenMP distribute to MLIR

### PR DESCRIPTION
**Pull request by skatrak**
This patch adds support for lowering the OpenMP DISTRIBUTE directive from PFT to MLIR. It only supports standalone DISTRIBUTE, support for composite constructs will come in follow-up PRs.

### CODE REVIEW

**Comment by skatrak** [comment]
This PR can now be reviewed, since #67720 with support for the `omp.distribute` MLIR operation was merged.

File: flang/lib/Lower/OpenMP.cpp [hunk 1]
```diff
@@ -1490,6 +1493,19 @@ bool ClauseProcessor::processDeviceType(
...
 }
 
+bool ClauseProcessor::processDistSchedule(
+    Fortran::lower::StatementContext &stmtCtx, mlir::UnitAttr &scheduleStatic,
+    mlir::Value &chunkSize) const {
+  if (auto *distScheduleClause = findUniqueClause<ClauseTy::DistSchedule>()) {
+    scheduleStatic = converter.getFirOpBuilder().getUnitAttr();
+    if (const auto *expr = Fortran::semantics::GetExpr(distScheduleClause->v)) {
+      chunkSize = fir::getBase(converter.genExprValue(*expr, stmtCtx));
+    }
```
Comment (TIFitis) [inline flang/lib/Lower/OpenMP.cpp:1503]: Nit: Remove braces
```suggestion
    if (const auto *expr = Fortran::semantics::GetExpr(distScheduleClause->v))
      chunkSize = fir::getBase(converter.genExprValue(*expr, stmtCtx));
```

**Review by TIFitis — APPROVED** [review]
LGTM :)

File: flang/lib/Lower/OpenMP.cpp [hunk 2]
```diff
@@ -1490,6 +1493,19 @@ bool ClauseProcessor::processDeviceType(
   return false;
 }
 
+bool ClauseProcessor::processDistSchedule(
+    Fortran::lower::StatementContext &stmtCtx, mlir::UnitAttr &scheduleStatic,
+    mlir::Value &chunkSize) const {
+  if (auto *distScheduleClause = findUniqueClause<ClauseTy::DistSchedule>()) {
```
Comment (TIFitis) [inline flang/lib/Lower/OpenMP.cpp:1499]: Nit: Expand `auto` here and below.

File: flang/lib/Lower/OpenMP.cpp (hunk 1 above)
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp]: Done

File: flang/lib/Lower/OpenMP.cpp (hunk 2 above)
Comment (skatrak) [reply flang/lib/Lower/OpenMP.cpp:1499]: The returned type is spelled out in the initialization expression here and all variables obtained from calling `Fortran::semantics::GetExpr` in this file use `auto` like below, so I prefer following the same approach.

**Review by DominikAdamski — APPROVED** [review]
LGTM. Please update tests before merge

**Review by kiranchandramohan — CHANGES_REQUESTED** [review]
This needs some more discussion or explanation. Could you construct a table of the lowerings for the various distribute combinations? Could you also mark if any of these are composite constructs?

**Comment by skatrak** [comment]
> This needs some more discussion or explanation. Could you construct a table of the lowerings for the various distribute combinations? Could you also mark if any of these are composite constructs?

The different constructs where DISTRIBUTE appears at the top, using the 5.0 spec (sect 2.9.4) as reference, are listed in the table below. All these can appear combined with TEAMS and TARGET, in which case they would be represented the same way but nested inside in the following manner:

```
!$omp teams distribute ...

omp.teams {
  omp.distribute {
    ...
  }
  omp.terminator
}
```

```
!$omp target teams distribute ...

omp.target {
  omp.teams {
    omp.distribute {
      ...
    }
    omp.terminator
  }
  omp.terminator
}
```

<table>
<tr>
<td>

**Construct**

</td>
<td>

**Composite**

</td>
<td>

**MLIR**

</td>
</tr>
<tr>
<td>

`!$omp distribute`

</td>
<td> No </td>
<td>

```mlir
omp.distribute {
  omp.wsloop ... {
    ...
  }
  omp.terminator
}
```

</td>
</tr>
<tr>
<td>

`!$omp distribute parallel do`

</td>
<td> Yes </td>
<td>

```mlir
omp.distribute {
  omp.parallel {
    omp.wsloop ... {
      ...
    }
    omp.terminator
  }
  omp.terminator
}
```

</td>
</tr>
<tr>
<td>

`!$omp distribute parallel do simd`

</td>
<td> Yes </td>
<td>

```mlir
omp.distribute {
  omp.parallel {
    omp.wsloop ... {
      <omp.simd?>
      ...
    }
    omp.terminator
  }
  omp.terminator
}
```

</td>
</tr>
<tr>
<td>

`!$omp distribute simd`

</td>
<td> Yes </td>
<td>

```mlir
omp.distribute {
  omp.simdloop ... {
    ...
  }
  omp.terminator
}
```

</td>
</tr>
</table>

**Comment by DominikAdamski** [comment]
PR: https://github.com/llvm/llvm-project/pull/79843 relates to lowering of `omp distributes` clause.

**Comment by skatrak** [comment]
This patch has now been updated to follow the recently introduced loop wrapper approach and is ready for review again. It only addresses standalone DISTRIBUTE and support for composite constructs will be left for follow-up PRs.

**Comment by skatrak** [comment]
Ping for review! Just updated the PR and addressed some conflicts with changes which landed recently.

File: flang/test/Lower/OpenMP/distribute.f90 [hunk 3]
```diff
@@ -0,0 +1,114 @@
+! REQUIRES: openmp_runtime
```
Comment (tblah) [inline flang/test/Lower/OpenMP/distribute.f90:1]: Why does this require the runtime?

Comment (skatrak) [reply flang/test/Lower/OpenMP/distribute.f90:1]: One of the functions contains an ALLOCATE clause that uses `omp_high_bw_mem_alloc` from `omp_lib`, which I think needs the openmp_runtime to work if I'm not wrong.

**Review by kiranchandramohan — APPROVED** [review]
Some of the tests do not check for omp.loop_nest (in loop-combined.f90). Is that missed accidentally or by design?

**Comment by kiranchandramohan** [comment]
> > Some of the tests do not check for omp.loop_nest (in loop-combined.f90). Is that missed accidentally or by design?
> 
> Thank you Kiran for giving this another look. In all these cases an `omp.loop_nest` is expected, but that test mainly just checks loop wrappers, which is also the case for if-clause.f90. The thinking behind it is that the existence of a single loop operation closely nested inside all loop wrappers is tested for each of the wrappers separately (distribute.f90, wsloop*.f90, etc) so I didn't think it to be necessary for combined/composite construct tests.
> 
> I agree it doesn't hurt to add these checks though, but maybe it's something I could do as a follow-up PR to update both tests rather than making unrelated test changes here. Let me know if that works for you.

That is fine.

**Pull request merged**